File: 1. Creational Design Patterns/Prototype/Python/PrototypeExample.py
# Prototype Pattern Example in Python
# Shape cloning system with different shape types

import copy
from abc import ABC, abstractmethod

class Shape(ABC):
    def __init__(self, x=0, y=0, color="white"):
        self.x = x
        self.y = y
        self.color = color
    
    @abstractmethod
    def clone(self):
        pass
    
    @abstractmethod
    def draw(self):
        pass

class Rectangle(Shape):
    def __init__(self, x=0, y=0, color="white", width=10, height=5):
        super().__init__(x, y, color)
        self.width = width
        self.height = height
    
    def clone(self):
        return copy.deepcopy(self)
    
    def draw(self):
        return f"Rectangle at ({self.x},{self.y}), color: {self.color}, size: {self.width}x{self.height}"

class Circle(Shape):
    def __init__(self, x=0, y=0, color="white", radius=5):
        super().__init__(x, y, color)
        self.radius = radius
    
    def clone(self):
        return copy.deepcopy(self)
    
    def draw(self):
        return f"Circle at ({self.x},{self.y}), color: {self.color}, radius: {self.radius}"

class ShapeRegistry:
    def __init__(self):
        self._prototypes = {}
    
    def register_prototype(self, name, prototype):
        self._prototypes[name] = prototype
    
    def create_shape(self, name):
        if name in self._prototypes:
            return self._prototypes[name].clone()
        return None

def main():
    # Create prototypes
    registry = ShapeRegistry()
    registry.register_prototype("blue_rectangle", Rectangle(0, 0, "blue", 20, 10))
    registry.register_prototype("red_circle", Circle(0, 0, "red", 15))
    
    # Clone and modify
    rect1 = registry.create_shape("blue_rectangle")
    rect1.x, rect1.y = 10, 20
    print(rect1.draw())
    
    rect2 = registry.create_shape("blue_rectangle")
    rect2.x, rect2.y = 30, 40
    rect2.width = 25
    print(rect2.draw())
    
    circle1 = registry.create_shape("red_circle")
    circle1.x, circle1.y = 50, 60
    print(circle1.draw())
    
    circle2 = registry.create_shape("red_circle")
    circle2.color = "green"
    circle2.radius = 20
    print(circle2.draw())

if __name__ == "__main__":
    main()

File: 1. Creational Design Patterns/Prototype/Typescript/PrototypeExample.ts
// Prototype Pattern – TypeScript Example
interface Shape {
  clone(): Shape;
  draw(): void;
}

class Rectangle implements Shape {
  constructor(
    public x = 0,
    public y = 0,
    public width = 0,
    public height = 0,
    public color = ''
  ) {}
  clone(): Shape {
    return new Rectangle(this.x, this.y, this.width, this.height, this.color);
  }
  draw(): void {
    console.log(`Rectangle: ${this.x},${this.y},${this.width},${this.height} color=${this.color}`);
  }
}

class Circle implements Shape {
  constructor(
    public x = 0,
    public y = 0,
    public radius = 0,
    public color = ''
  ) {}
  clone(): Shape {
    return new Circle(this.x, this.y, this.radius, this.color);
  }
  draw(): void {
    console.log(`Circle: ${this.x},${this.y},r=${this.radius} color=${this.color}`);
  }
}

const rect1 = new Rectangle(10, 20, 30, 40, 'red');
const rect2 = rect1.clone() as Rectangle;
rect2.x += 100;
rect1.draw();
rect2.draw();
